### DISCUSSION TITLE
how to distribute 64bit C++ dll file

### DISCUSSION THREAD

**Discussion by leahwang1962** [wixtoolset/issues#7627] [Questions]
**Encountered at:** wixtoolset/repo@sha

Hi,
I am using a 64 bit C++ dll file in my VB.net program. As shown in the attached product.wxs, the dll file will be deployed in the INSTALLFOLDER folder with exe file. After run the msi setup file, the dll and exe files are indeed deployed under C:\Program Files <x86>\. But when I run the program, I get error "unable to load dll file". This error only happens at computer without VS.net C++ installed.  
Thanks for help.

<img width="895" alt="productwxs" src="https://github.com/wixtoolset/issues/assets/99756010/8bd30900-53fc-4b91-8888-65a365123d97">
<img width="895" alt="dll file in program file folder" src="https://github.com/wixtoolset/issues/assets/99756010/1824a4ca-df36-462a-ae93-3f6bb61398c5">
<img width="895" alt="Notfound dll file" src="https://github.com/wixtoolset/issues/assets/99756010/cc1d7acc-53aa-4f17-8192-cacdb1028c2d">

**Comment by barnson** [comment]
WiX doesn't use issues for support. We moved this issue to the [GitHub discussions area][gh]. You might also consider asking your question on [Stack Overflow][so].

[gh]: https://github.com/wixtoolset/issues/discussions?discussions_q=category%3AQuestions
[so]: https://stackoverflow.com/questions/tagged/wix

**Comment by pas059** [comment]
after seeing the second view, your dll & exe are installed in the Program Files (x86) folder, so as a 32 bits application; consequently, Windows expects to find only 32 bits components when you launch your app and not a 64 bits dll

**Reply by leahwang1962** [reply]
Hi pas059,
Thanks for your reply. 
Following your opinion, I built a 32bit C++ dll( it is OK for my application), but the error are the same as before. Please see the screen shot as how the C++ dll is built and the error message. Please note, my app after deployment works fine on computer installed with Visual studio c++. But if I deploy it on other computer without VS C++, it will show the dll can't found.
<img width="895" alt="32bitdllbuilt" src="https://github.com/wixtoolset/issues/assets/99756010/2ebe5e95-58c0-4094-9539-950cf5d83116">
![32bitdll_can'tfound](https://github.com/wixtoolset/issues/assets/99756010/41fec6e8-883c-488e-9e7d-380064ce8451)

**Reply by BMurri** [reply]
> Please note, my app after deployment works fine on computer installed with Visual studio c++. But if I deploy it on other computer without VS C++, it will show the dll can't found.

This suggests that you linked your C++ to the dynamic runtime instead of the static runtime. This is a build setting in your C++ project.

The reason is that when loading DLLs, if any must-load dependency (all the way up the chain) cannot be found, a not-found error for the requested DLL is reported.

**Reply by leahwang1962** [reply]
Hi BMurri,
Thanks for your response.
Would you please elaborate the static runtime: how to build it in C++, and how to import the functions to VB.net?
In C++ property page as attached here, I found 2 locations related to static runtime, but I am not sure which one I should use: Static library( to build a .lib file) or Multithread DLL/MD(to build a .dll file). 
Thanks a lot.
<img width="895" alt="StaticDll" src="https://github.com/wixtoolset/issues/assets/99756010/3aea6f9b-7d63-4f11-9586-125678568df9">

**Comment by pas059** [comment]
hi,
the second. As i know static lib can not be used in .net code, only dll using DllImport

**Reply by leahwang1962** [reply]
Hi Pas059,
Appreciate your help.
As shown in the screen shot, unfortunately I have to include a static  libfli.lib file in my C++ dll.
After changing to multi-thread Dll (/MD) runtime library, I found it is true this libfli.lib file causing the "unable to load dll " error because when I removed it from C++ dll, the distributed program runs fine on any computer.

Any suggestion how to handle this libfli.lib file in my C++ dll?
 
<img width="895" alt="LibFile" src="https://github.com/wixtoolset/issues/assets/99756010/62585dfd-7572-47c5-a3f3-8b1f40c5dd9c">

**Reply by BMurri** [reply]
When building any C++ code, a `lib` file is incorporated into the binary created in your build, so those files do not require redistribution. Referenced `dll` files generally do need to be redistributed unless they are platform resident.

The normal fix for the original error is to select a runtime setting that does NOT contain "DLL".

**Reply by leahwang1962** [reply]
Thank you BMurri. 
After I added the Referenced dll related with .lib in the distribution package, the error is gone.
